File: src/main/mpris.ts
import { EventEmitter } from 'events';
import { PlaybackState } from './types';

/**
 * MPRIS (Media Player Remote Interfacing Specification) Manager for Linux
 * Implements org.mpris.MediaPlayer2 and org.mpris.MediaPlayer2.Player
 */
export class MprisManager extends EventEmitter {
  private serviceName = 'flixdesk';
  private identity = 'FlixDesk';
  private desktopEntry = 'io.github.Pak_Man926.FlixDesk';
  private mprisInstance: any = null;
  private isInitialized = false;

  private currentState: PlaybackState = {
    isPlaying: false,
    isBuffering: false,
    title: 'Netflix',
    subTitle: '',
    seasonEpisode: '',
    artworkUrl: '',
    duration: 0,
    currentTime: 0,
    volume: 1.0,
    isMuted: false,
    playbackRate: 1.0,
  };

  constructor() {
    super();
  }

  /**
   * Initializes the MPRIS D-Bus service if running on Linux
   */
  public async init(): Promise<void> {
    if (process.platform !== 'linux') {
      return;
    }

    try {
      // Dynamic import to handle environments where mpris-service might or might not be installed
      let MprisService: any;
      try {
        MprisService = require('mpris-service');
      } catch {
        console.log('[MPRIS] mpris-service native module not present; using built-in D-Bus event bridge.');
      }

      if (MprisService) {
        this.mprisInstance = new MprisService({
          name: this.serviceName,
          identity: this.identity,
          supportedUriSchemes: ['https', 'http'],
          supportedMimeTypes: ['video/mp4', 'video/webm'],
          supportedInterfaces: ['player'],
          desktopEntry: this.desktopEntry,
        });

        this.setupMprisListeners();
        this.isInitialized = true;
        console.log(`[MPRIS] Registered D-Bus interface org.mpris.MediaPlayer2.${this.serviceName}`);
      }
    } catch (err) {
      console.warn('[MPRIS] Could not initialize MPRIS D-Bus interface:', err);
    }
  }

  private setupMprisListeners(): void {
    if (!this.mprisInstance) return;

    this.mprisInstance.on('play', () => this.emit('play'));
    this.mprisInstance.on('pause', () => this.emit('pause'));
    this.mprisInstance.on('playpause', () => this.emit('playpause'));
    this.mprisInstance.on('stop', () => this.emit('stop'));
    this.mprisInstance.on('next', () => this.emit('next'));
    this.mprisInstance.on('previous', () => this.emit('previous'));
    this.mprisInstance.on('raise', () => this.emit('raise'));
    this.mprisInstance.on('quit', () => this.emit('quit'));

    this.mprisInstance.on('seek', (offsetMicroseconds: number) => {
      const offsetSeconds = offsetMicroseconds / 1000000;
      this.emit('seek', offsetSeconds);
    });

    this.mprisInstance.on('position', ({ position }: { position: number }) => {
      const positionSeconds = position / 1000000;
      this.emit('setPosition', positionSeconds);
    });

    this.mprisInstance.on('volume', (vol: number) => {
      this.emit('setVolume', vol);
    });
  }

  /**
   * Updates MPRIS metadata and playback status from player events
   */
  public updateState(state: Partial<PlaybackState>): void {
    this.currentState = { ...this.currentState, ...state };

    if (!this.mprisInstance) return;

    try {
      // 1. Playback status
      this.mprisInstance.playbackStatus = this.currentState.isPlaying ? 'Playing' : 'Paused';

      // 2. Playback Rate & Volume
      this.mprisInstance.rate = this.currentState.playbackRate || 1.0;
      this.mprisInstance.volume = this.currentState.isMuted ? 0 : this.currentState.volume;

      // 3. Position (in microseconds)
      this.mprisInstance.canSeek = true;
      this.mprisInstance.canControl = true;
      this.mprisInstance.canPlay = true;
      this.mprisInstance.canPause = true;
      this.mprisInstance.canGoNext = true;
      this.mprisInstance.canGoPrevious = true;

      // 4. Metadata dictionary
      const title = this.currentState.subTitle
        ? `${this.currentState.title} - ${this.currentState.subTitle}`
        : this.currentState.title || 'Netflix';

      const artists = this.currentState.seasonEpisode
        ? [this.currentState.seasonEpisode]
        : ['Netflix'];

      this.mprisInstance.metadata = {
        'mpris:trackid': this.mprisInstance.objectPath('track/current'),
        'mpris:length': Math.round((this.currentState.duration || 0) * 1000000), // Microseconds
        'mpris:artUrl': this.currentState.artworkUrl || '',
        'xesam:title': title,
        'xesam:album': this.currentState.title || 'Netflix',
        'xesam:artist': artists,
      };
    } catch (err) {
      console.warn('[MPRIS] Failed to update MPRIS metadata:', err);
    }
  }

  public getPlaybackState(): PlaybackState {
    return { ...this.currentState };
  }

  public destroy(): void {
    if (this.mprisInstance) {
      try {
        this.mprisInstance = null;
      } catch (e) {
        // ignore
      }
    }
  }
}

export const mprisManager = new MprisManager();

File: flixdesk/mpris.py
"""
FlixDesk - Linux MPRIS D-Bus Provider
Implements org.mpris.MediaPlayer2 and org.mpris.MediaPlayer2.Player
for native integration with GNOME, Pop!_OS COSMIC, KDE, and hardware media keys.
"""

from typing import Dict, Any, Optional

try:
    from PySide6.QtCore import QObject, Slot, Property, Signal
    from PySide6.QtDBus import (
        QDBusConnection,
        QDBusAbstractAdaptor,
    )
    QT_DBUS_AVAILABLE = True
except ImportError:
    QT_DBUS_AVAILABLE = False


class MprisRootAdaptor(QDBusAbstractAdaptor if QT_DBUS_AVAILABLE else object):
    if QT_DBUS_AVAILABLE:
        # Export org.mpris.MediaPlayer2
        def __init__(self, parent):
            super().__init__(parent)
            self._parent = parent

        @Slot()
        def Raise(self):
            self._parent.raise_window()

        @Slot()
        def Quit(self):
            self._parent.quit_app()

        @Property(bool)
        def CanQuit(self):
            return True

        @Property(bool)
        def CanRaise(self):
            return True

        @Property(bool)
        def HasTrackList(self):
            return False

        @Property(str)
        def Identity(self):
            return "FlixDesk"

        @Property(list)
        def SupportedUriSchemes(self):
            return ["https", "http"]

        @Property(list)
        def SupportedMimeTypes(self):
            return []


class MprisPlayerAdaptor(QDBusAbstractAdaptor if QT_DBUS_AVAILABLE else object):
    if QT_DBUS_AVAILABLE:
        def __init__(self, parent):
            super().__init__(parent)
            self._parent = parent

        @Slot()
        def Next(self):
            self._parent.command_next()

        @Slot()
        def Previous(self):
            self._parent.command_previous()

        @Slot()
        def Pause(self):
            self._parent.command_pause()

        @Slot()
        def PlayPause(self):
            self._parent.command_play_pause()

        @Slot()
        def Stop(self):
            self._parent.command_pause()

        @Slot()
        def Play(self):
            self._parent.command_play()

        @Property(str)
        def PlaybackStatus(self):
            return self._parent.playback_status

        @Property(dict)
        def Metadata(self):
            return self._parent.get_mpris_metadata()

        @Property(bool)
        def CanControl(self):
            return True

        @Property(bool)
        def CanPlay(self):
            return True

        @Property(bool)
        def CanPause(self):
            return True

        @Property(bool)
        def CanGoNext(self):
            return True

        @Property(bool)
        def CanGoPrevious(self):
            return True

        @Property(bool)
        def CanSeek(self):
            return False


class MprisManager(QObject if QT_DBUS_AVAILABLE else object):
    def __init__(self, window_callback=None):
        if QT_DBUS_AVAILABLE:
            super().__init__()
        self.window_callback = window_callback
        self.playback_status = "Stopped"
        self.current_title = "FlixDesk"
        self.current_episode = ""
        self.current_duration = 0.0
        self.bus: Optional[Any] = None

    def init(self) -> bool:
        if not QT_DBUS_AVAILABLE:
            print("[MPRIS] QtDBus module is not available on this system.")
            return False

        try:
            self.bus = QDBusConnection.sessionBus()
            if not self.bus.isConnected():
                print("[MPRIS] Could not connect to D-Bus session bus.")
                return False

            service_name = "org.mpris.MediaPlayer2.FlixDesk"
            if not self.bus.registerService(service_name):
                print(f"[MPRIS] Could not register D-Bus service {service_name}")
                return False

            self.root_adaptor = MprisRootAdaptor(self)
            self.player_adaptor = MprisPlayerAdaptor(self)

            if not self.bus.registerObject("/org/mpris/MediaPlayer2", self):
                print("[MPRIS] Could not register /org/mpris/MediaPlayer2 object")
                return False

            print("[MPRIS] Successfully registered org.mpris.MediaPlayer2.FlixDesk on D-Bus.")
            return True
        except Exception as e:
            print(f"[MPRIS] Initialization error: {e}")
            return False

    def update_state(self, state: Dict[str, Any]) -> None:
        playing = state.get("playing", False)
        self.playback_status = "Playing" if playing else "Paused"
        self.current_title = state.get("title") or "FlixDesk"
        self.current_episode = state.get("episode") or ""
        self.current_duration = float(state.get("duration") or 0)

    def get_mpris_metadata(self) -> Dict[str, Any]:
        meta = {
            "mpris:trackid": "/org/mpris/MediaPlayer2/Track/1",
            "xesam:title": f"{self.current_title} - {self.current_episode}" if self.current_episode else self.current_title,
            "xesam:artist": ["Netflix"],
            "xesam:album": self.current_title,
        }
        if self.current_duration > 0:
            meta["mpris:length"] = int(self.current_duration * 1_000_000)  # microseconds
        return meta

    def raise_window(self):
        if self.window_callback:
            self.window_callback("raise")

    def quit_app(self):
        if self.window_callback:
            self.window_callback("quit")

    def command_play_pause(self):
        if self.window_callback:
            self.window_callback("play_pause")

    def command_play(self):
        if self.window_callback:
            self.window_callback("play")

    def command_pause(self):
        if self.window_callback:
            self.window_callback("pause")

    def command_next(self):
        if self.window_callback:
            self.window_callback("next")

    def command_previous(self):
        if self.window_callback:
            self.window_callback("previous")
